### PR TITLE
CORS-3546: Nutanix: add gpus and dataDisks support

### PR DESCRIPTION
**Pull request by yanhua121**
Update with the api changes for supporting Nutanix gpus and dataDisks:
https://github.com/openshift/api/pull/1935

Add Machine webhooks to validate the newly added Nutanix providerSpec fields.

Manual in-lab testing of the this PR passed.

### CODE REVIEW

File: pkg/webhooks/machine_webhook.go [hunk 1]
```diff
@@ -1665,12 +1665,121 @@ func validateNutanix(m *machinev1beta1.Machine, config *admissionConfig) (bool,
...
+		// validate the minimum diskSize of 1Gi
+		diskSizeBytes := disk.DiskSize.Value()
+		if diskSizeBytes < 1024*1024*1024 {
+			fldErrs = append(fldErrs, field.Invalid(fldPath.Child("diskSize"), fmt.Sprintf("%v bytes", diskSizeBytes), "The minimum diskSize is 1Gi bytes."))
+		}
+
+		if disk.DeviceProperties != nil {
+			switch disk.DeviceProperties.DeviceType {
+			case machinev1.NutanixDiskDeviceTypeDisk:
+				if disk.DeviceProperties.AdapterType != machinev1.NutanixDiskAdapterTypeSCSI &&
```
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:1715]: You'd probably be able to make this easier to read if you used a switch statement here and had the default case as the error case

File: pkg/webhooks/machine_webhook.go [hunk 2]
```diff
@@ -1665,12 +1665,121 @@ func validateNutanix(m *machinev1beta1.Machine, config *admissionConfig) (bool,
...
+					disk.DeviceProperties.AdapterType != machinev1.NutanixDiskAdapterTypePCI &&
+					disk.DeviceProperties.AdapterType != machinev1.NutanixDiskAdapterTypeSATA &&
+					disk.DeviceProperties.AdapterType != machinev1.NutanixDiskAdapterTypeSPAPR {
+					fldErrs = append(fldErrs, field.Invalid(fldPath.Child("deviceProperties", "adapterType"), disk.DeviceProperties.AdapterType,
+						fmt.Sprintf("invalid adapter type for the %q device type, the valid values: %q, %q, %q, %q, %q.",
+							machinev1.NutanixDiskDeviceTypeDisk, machinev1.NutanixDiskAdapterTypeSCSI, machinev1.NutanixDiskAdapterTypeIDE,
+							machinev1.NutanixDiskAdapterTypePCI, machinev1.NutanixDiskAdapterTypeSATA, machinev1.NutanixDiskAdapterTypeSPAPR)))
+				}
+			case machinev1.NutanixDiskDeviceTypeCDROM:
+				if disk.DeviceProperties.AdapterType != machinev1.NutanixDiskAdapterTypeIDE &&
```
Comment (JoelSpeed) [inline pkg/webhooks/machine_webhook.go:1726]: As above, maybe use a switch?

File: pkg/webhooks/machine_webhook.go (hunk 1 above)
Comment (yanhua121) [reply pkg/webhooks/machine_webhook.go]: Changed as suggested.

**Review by elmiko — COMMENTED** [review]
looks mostly fine to me, but i'm not sure how the `fldErrs` variable in the disk validator is being declared.

File: pkg/webhooks/machine_webhook.go [hunk 3]
```diff
@@ -1665,12 +1665,124 @@ func validateNutanix(m *machinev1beta1.Machine, config *admissionConfig) (bool,
...
 
+func validateNutanixDataDisks(disks []machinev1.NutanixVMDisk) (fldErrs []*field.Error) {
+	fldPath := field.NewPath("providerSpec", "dataDisks")
+	var errMsg string
+
+	for _, disk := range disks {
+		// validate the minimum diskSize of 1Gi
+		diskSizeBytes := disk.DiskSize.Value()
+		if diskSizeBytes < 1024*1024*1024 {
+			fldErrs = append(fldErrs, field.Invalid(fldPath.Child("diskSize"), fmt.Sprintf("%v bytes", diskSizeBytes), "The minimum diskSize is 1Gi bytes."))
```
[the remark this answers is not in the document]
Comment (elmiko) [reply pkg/webhooks/machine_webhook.go:1709]: oops, i missed the declared name in the return spec for the function

**Review by elmiko — COMMENTED** [review]
/lgtm

**Comment by elmiko** [comment]
i'd like to see the e2e-nutanix test pass, and also get another review on this

**Review by elmiko — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**